bee-api/bee-rest-api/src/types/dtos.rs

```diff
@@ -0,0 +1,954 @@
+// Copyright 2020 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::types::error::Error;
+
+use bee_ledger::types::Receipt;
+use bee_message::{
+    address::{Address, Ed25519Address, ED25519_ADDRESS_LENGTH},
+    input::{Input, TreasuryInput, UtxoInput},
+    milestone::MilestoneIndex,
+    output::{Output, SignatureLockedDustAllowanceOutput, SignatureLockedSingleOutput, TreasuryOutput},
+    parents::Parents,
+    payload::{
+        indexation::IndexationPayload,
+        milestone::{
+            MilestoneId, MilestonePayload, MilestonePayloadEssence, MILESTONE_MERKLE_PROOF_LENGTH,
+            MILESTONE_PUBLIC_KEY_LENGTH,
+        },
+        receipt::{MigratedFundsEntry, ReceiptPayload, TailTransactionHash, TAIL_TRANSACTION_HASH_LEN},
+        transaction::{Essence, RegularEssence, TransactionId, TransactionPayload},
+        treasury::TreasuryTransactionPayload,
+        Payload,
+    },
+    signature::{Ed25519Signature, SignatureUnlock},
+    unlock::{ReferenceUnlock, UnlockBlock, UnlockBlocks},
+    Message, MessageBuilder, MessageId,
+};
+use bee_protocol::types::peer::Peer;
+
+use serde::{Deserialize, Serialize, Serializer};
+use serde_json::Value;
+
+use std::convert::{TryFrom, TryInto};
+
+/// The message object that nodes gossip around in the network.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct MessageDto {
+    #[serde(rename = "networkId")]
+    pub network_id: String,
+    #[serde(rename = "parentMessageIds")]
+    pub parents: Vec<String>,
+    pub payload: Option<PayloadDto>,
+    pub nonce: String,
+}
+
+impl From<&Message> for MessageDto {
+    fn from(value: &Message) -> Self {
+        MessageDto {
+            network_id: value.network_id().to_string(),
+            parents: value.parents().iter().map(|p| p.to_string()).collect(),
+            payload: value.payload().as_ref().map(Into::into),
+            nonce: value.nonce().to_string(),
+        }
+    }
+}
```

Comment:
Are those `Dto` reimplementations of bee-message types really unavoidable? It seems like that a change in either `bee-message` or `bee-api` results in a version bump of the respective other crate.

Reply:
DTOs are usually considered best practice for many reasons: https://stackoverflow.com/a/36175349/8363666

Reply:
I am not arguing against DTOs in general. But in our case it seems like we violate the DRY principle. That weren't too bad a thing if it would be contained in one crate. But here we create a very tight coupling between `bee-message` and `bee-api` causing one to version bump if the other one does.